### PR TITLE
feat: error handler for trpc

### PR DESCRIPTION
**Pull request by ted-pulsen**
Closes #<issue>

## ✅ Checklist

- [x] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-08-15).
- [x] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] I performed a functional test on my final commit

---

## Changelog

Added an error handler for tRPC.

---

## Screenshots

Upon hitting the endpoint with an invalid request model
```bash
$ curl --get http://localhost:3000/api/trpc/example.hello \
  --data-urlencode "batch=1" \
  --data-urlencode 'input={"0":{"json":{"text":42}}}'
```
image

💯

### CODE REVIEW

**Comment by c-ehrlich** [comment]
this is a good idea but maybe it should only be on in development?

**Comment by ted-pulsen** [comment]
I agree, good point

**Comment by ted-pulsen** [comment]
Another option is to move the if statement outside of `onError`.

```typescript
  onError: process.env.NODE_ENV === "development"
    ? ({ path, error }) => {
        console.error(`❌ tRPC failed on ${path}: ${error}`);
      }
    : undefined,
```

Pros:
- In production, it won't run an empty error handler.
- In development, it won't revalidate that it's still in development.

Cons:
- It's less obvious what's going on.
- Might encourage code duplication if similar logging is performed.

**Comment by juliusmarminge** [comment]
Is it really necessary to have this since we have the loggerlink?

**Comment by ted-pulsen** [comment]
I notice that other logs (from next.js I presume) are written in a different format and coloring. However, I can't find where to import it so that loggerlink can be reused here.

File: cli/template/addons/trpc/api-handler.ts [hunk 1]
```diff
@@ -7,4 +7,10 @@ import { createContext } from "../../../server/router/context";
 export default createNextApiHandler({
   router: appRouter,
   createContext,
+  onError:
+    process.env.NODE_ENV === "development"
```
[the remark this answers is not in the document]
Comment (ted-pulsen) [reply cli/template/addons/trpc/api-handler.ts:11]: It would be cleaner but that would evaluate to a boolean rather than the function needed here.

**Comment by juliusmarminge** [comment]
> I notice that other logs (from next.js I presume) are written in a different format and coloring. However, I can't find where to import it so that loggerlink can be reused here.

I'm just saying that we are already logging tRPC requests (all reqs. in dev and errors in prod) in the loggerlink:

https://github.com/t3-oss/create-t3-app/blob/ff64216a89261474dee966272a8e49ffd0b1b9a6/cli/template/page-studs/_app/with-trpc.tsx#L30-L34

I understand that this log is explicitly for errors but is it really necessary to have both?

File: cli/template/addons/trpc/api-handler.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (juliusmarminge) [reply cli/template/addons/trpc/api-handler.ts:11]: Yea, sometimes it works and sometimes it doesn't. Not a big deal

**Comment by ted-pulsen** [comment]
loggerLink handles logging of tRPC requests on the client but, at least from what I've seen those logs are not visible server-side. With `onError` you'll be reported on malformed or otherwise failed requests, these logs can then be collected (via `docker logs` or other) for centralized (via `Elasticsearch` or other) analyses.

**Comment by ted-pulsen** [comment]
I've validated that links are only run client-side, try adding a custom link and you'll find that it's logging only on the client.

```typescript
      links: [
        function customLink() {
          return ({ prev, next, op }) => {
            next(op, (result) => {
              console.log('*** FOOBAR ***');
              prev(result);
            });
          }
        },
        loggerLink({
          enabled: (opts) =>
            process.env.NODE_ENV === "development" ||
            (opts.direction === "down" && opts.result instanceof Error),
          
        }),
        httpBatchLink({ url }),
      ],
```

**Comment by juliusmarminge** [comment]
Aight sounds good then. We should probably use our validated env-vars though (even though NODE_ENV is a default one that's always there but still)?

**Comment by ted-pulsen** [comment]
Nice feature, I've not noticed those env-vars before.

File: cli/template/addons/trpc/api-handler.ts [hunk 2]
```diff
@@ -8,7 +9,7 @@ export default createNextApiHandler({
   router: appRouter,
   createContext,
   onError:
-    process.env.NODE_ENV === "development"
+    env.NODE_ENV === "development"
```
[the remark this answers is not in the document]
Comment (ted-pulsen) [reply cli/template/addons/trpc/api-handler.ts:12]: I'm adding it as we speak, soon an update

[the remark this answers is not in the document]
Comment (juliusmarminge) [reply cli/template/addons/trpc/api-handler.ts:12]: I'm pretty positive we've had it at some point, no?

[the remark this answers is not in the document]
Comment (juliusmarminge) [reply cli/template/addons/trpc/api-handler.ts:12]: It's in the two prisma schemas, it should be in all of them (including base)

https://github.com/t3-oss/create-t3-app/blob/ff64216a89261474dee966272a8e49ffd0b1b9a6/cli/template/addons/env/auth-prisma-schema.mjs#L10

https://github.com/t3-oss/create-t3-app/blob/ff64216a89261474dee966272a8e49ffd0b1b9a6/cli/template/addons/env/prisma-schema.mjs#L10

File: cli/src/installers/envVars.ts [hunk 3]
```diff
@@ -9,6 +9,7 @@ export const envVariablesInstaller: Installer = async ({
 }) => {
   const usingAuth = packages?.nextAuth.inUse;
   const usingPrisma = packages?.prisma.inUse;
+  const usingTrpc = packages?.trpc.inUse;
```
Comment (juliusmarminge) [inline cli/src/installers/envVars.ts:12]: this is not needed. we have a default schema in template/base which we should use instead.

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
we do not need a separate trpc schema, just include NODE_ENV in

https://github.com/t3-oss/create-t3-app/blob/ff64216a89261474dee966272a8e49ffd0b1b9a6/cli/template/base/src/env/schema.mjs#L8

and

https://github.com/t3-oss/create-t3-app/blob/ff64216a89261474dee966272a8e49ffd0b1b9a6/cli/template/addons/env/auth-schema.mjs#L8

File: cli/src/installers/envVars.ts (hunk 3 above)
Comment (ted-pulsen) [reply cli/src/installers/envVars.ts:12]: Ok, I'll do that instead.

File: cli/template/base/src/env/schema.mjs [hunk 4]
```diff
@@ -6,7 +6,7 @@ import { z } from "zod";
  * This way you can ensure the app isn't built with invalid env vars.
  */
 export const serverSchema = z.object({
-  // SERVERVAR: z.string(),
+  NODE_ENV: z.enum(["development", "test", "production"]),
```
Comment (juliusmarminge) [inline cli/template/base/src/env/schema.mjs:9]: Nice.

We also need it in the auth-schema:

https://github.com/t3-oss/create-t3-app/blob/ff64216a89261474dee966272a8e49ffd0b1b9a6/cli/template/addons/env/auth-schema.mjs#L8-L13

Comment (ted-pulsen) [reply cli/template/base/src/env/schema.mjs:9]: Done

**Review by juliusmarminge — APPROVED** [review]
Thanks for the patience 🙏🏼

**Comment by ted-pulsen** [comment]
No problem, thanks for the guidance!

Any reason why the env-var stuff is not in typescript?

**Comment by juliusmarminge** [comment]
> Any reason why the env-var stuff is not in typescript?

We import it in the `next.config.mjs` to validate them at build-time and you can't import typescript-files in there.

**Pull request merged**